### PR TITLE
[MRG] add sqlite3 implementations for `Index`, `CollectionManifest`, and `LCA_Database`

### LINKED ISSUE #821: consider ways to improve speed of LCA database
Right now LCA databases are loaded into memory from a JSON file, but loading big JSON files (esp in Python) is slow and causes significant delay at startup. It'd be nice to have an on disk option like with SBTs.

I have been thinking about using something like a SQLite database to support the basic "hash -> idx" lookup, since that is the bulk of the data AFAIK. Or we could use some other on-disk dictionary structure, but of course that gets slow for big databases.

This could be implemented as an optional caching mechanism, too, but that could add a lot of nasty code.

### LINKED ISSUE #948: Oxidize parts of LCA_Database?
An implicit goal of #946 refactoring and testing the `LCA_Database` class is to support optimization of searches by moving them into a Rust extension module.  I think the function `_find_signatures(...)` would be the main target (along with `_signatures(...)`). Could be good to look at some of the other LCA submodule commands to see if the hashvalue gathering done in classify and summarize could be pulled back to an `LCA_Database` method as well.

### LINKED ISSUE #1111: use a faster `json` library in `lca/lca_db.py`
perhaps one reason LCA databases are so slow is that we're using a slow JSON parser - the stdlib one!

ref #948 #909 and oddly #737

### LINKED ISSUE #1807: storing signatures/sketches in a sqlite DB
For tangential reasons, I ended up writing storage and retrieval code for storing actual sketches (incl minhash values) in a SQLite database.

See https://github.com/ctb/2022-sourmash-sqlite.

It seems to work for flat sketches - the following roundtrip does the right thing, at any rate:
```
./save-mh-to-sqlite.py test-db/all.zip -o all.db
./load-mh-from-sqlite.py all.db -o all.zip
```
reproduces all the signatures from `test-db/all.zip` in `all.zip`, including protein etc.

I still need to add abundance tracking tho. Oops.

Will link in related issues later.

### LINKED ISSUE #1930: should we store signatures in SQLite databases?
In PR https://github.com/sourmash-bio/sourmash/pull/1808, I'm nearing a decision point.

This PR is an experimental PR that adds `SqliteCollectionManifest` and `SqliteIndex`, for storing manifests and signatures in a SQLite database. `SqliteCollectionManifest` is indepenently useful and can be wrapped in a `StandaloneManifestIndex`, while `SqliteIndex` builds on `SqliteCollectionManifest`.

My conundrum is that the manifest class is almost certainly worth merging, but I'm not sure about the index class. This issue is to track this discussion independently of the PR.

## Questions:

Two big questions arise:

### should we merge `SqliteIndex`?

`SqliteCollectionManifest` is really useful, but it's not clear to me that `SqliteIndex` should be merged.

arguments _for_ `SqliteIndex`:

* it works! working code is good!
* low startup overhead (benchmark!)
* low memory (benchmark!)
* multithread/multiprocess for read-only db
* another example of on-disk index
* efficient (?) on-the-fly rescaling - benchmark!

arguments _against_ `SqliteIndex`:

* more code to support
* code is reasonably complicated

###  should we support scaled=1 in `SqliteIndex`? 

It adds a dependency, makes the code more complicated, and is not that fast. On the other hand it's kind of neat and it seems to work ok :).

### PR DESCRIPTION
**Pull request by ctb**
This PR adds three new SQLite-based features to sourmash, `SqliteIndex`, `SqliteCollectionManifest`, and `LCA_SqliteDatabase`. These new features provide the following features (in order of their estimated importance to the future world domination of sourmash):

* fast SQLite-based manifest support for extremely large collections
* low-memory, low-latency, and multiprocess LCA databases for better LCAing,
* a more unified approach to SQLite database handling in sourmash,
* a reasonably nice alternative to other indexed collections;

Fixes https://github.com/sourmash-bio/sourmash/issues/1807
Fixes https://github.com/sourmash-bio/sourmash/issues/1930
Fixes https://github.com/sourmash-bio/sourmash/issues/948 by providing a faster `LCA_Database`
Fixes https://github.com/sourmash-bio/sourmash/issues/821
Fixes https://github.com/sourmash-bio/sourmash/issues/1111 by using SQLite rather than JSON.

Provides one solution for https://github.com/sourmash-bio/sourmash/issues/909, a shared LCA index for many processes.

## Highlights

**`SqliteIndex`** is a SQLite based Index class that:
* stores sketches and hashes in a sqlite db;
* is restricted to storing scaled sketches with a single scaled, albeit with multiple ksizes and moltypes;
* does efficient, on-the-fly downsampling of sketches;

**`SqliteCollectionManifest`** is a SQLite based Manifest class that:
* stores manifest entries in sqlite;
* handles selection and picklists appropriately, in SQL (so, fast!);
* joins `CollectionManifest` in being loadable via the CLI in a `StandaloneManifestIndex` as a first-class Index.

**`LCA_SqliteDatabase`** is a SQLite-based LCA class built on top of `SqliteIndex` and `LineageDB_Sqlite` that supports all of the LCA commands.

This PR introduces much of the remaining functionality from "manifests of manifests" https://github.com/sourmash-bio/sourmash/issues/1652 https://github.com/sourmash-bio/sourmash/issues/1685 into sourmash, after merge of `StandaloneManifestIndex` in #1891.

## Reading and writing

All classes are fully integrated into the sourmash command-line functionality.

Collections of all three types can be read as regular ol' `Index` classes using `sourmash.load_file_as_index(...)`.

SQLite databases can be output with `-o <name>.sqldb`, and are read natively on the command line.

SQLite manifests can be output by passing `-F sql` to `sourmash sig manifest`, and will be read natively by sourmash as `StandaloneManifestIndex`.

SQLite LCA databases can be output by passing `-F sql` to `sourmash lca index`.

## Other additions

This PR also:
* fixes GitHub workflow caching to track setup.cfg, and provides a version number to manually increment when new dependencies are added;
* adds the `-d/--debug` flag to `sourmash search`
* fixes a bug where `LCA_Database.signatures(...)` didn't pay attention to picklists.

## Implementation notes

`SqliteIndex` is built on `SqliteCollectionManifest`, and `SqliteCollectionManifest` is usable without `SqliteIndex`.

`SqliteIndex` and `SqliteCollectionManifest` share  the `sketches` table.

`SqliteIndex` supports a scaled of 1 - that is, direct storage of unsigned 64-bit numbers! - via the `bitstring` library, which converts unsigned long longs into signed long longs that sqlite can store. **This is a new dependency.**

## SQLite databases, and backwards compatibility

This PR also adds a required table to all sourmash SQLite databases: the table `sourmash_internal` contains key, value pairs that document what version of `SqliteIndex`, `SqliteManifest`, and `SqliteLineage` (taxonomy) are supported by the database. See the module `sourmash.sqlite_utils` for supporting code.

The only previous use of SQLite in sourmash was in taxonomy databases, for the `sourmash.tax` module, output by `sourmash tax prepare`. This code used the table `taxonomy`. Moving forward this table has been renamed to `sourmash_taxonomy` and taxonomy SQLite databases also contain the `sourmash_internal` table with a `SqliteLineage` entry. The previous use is fully supported by backwards compatibility code.

## Technical notes

(the below is also enshrined in the docstring in `index/sqlite_index.py`) -

```
sqlite3 based Index, CollectionManifest, and LCA_Database
implementations.

These classes support a variety of flexible and fast on-disk storage,
search, and retrieval functions.

SqliteIndex stores full scaled signatures; sketches are stored as
reverse-indexed collections of hashes. Search is optimized via the
reverse index. Num and abund sketches are not supported. All scaled
values must be the same upon insertion. Multiple moltypes _are_
supported.

SqliteCollectionManifest provides a full implementation of the
manifest API. It can store details for all signature types. When used
as part of a SqliteIndex database, it does not support independent
insertion.

LCA_SqliteDatabase builds on top of SqliteIndex and LineageDB_Sqlite
(in the tax submodule) to provide a full on-disk implementation of
LCA_Database.

Using these classes
-------------------

These classes are fully integrated into sourmash loading.
Internally, use `sqlite_index.load_sqlite_index(...)` to load a specific
file; this will return the appropriate SqliteIndex, StandaloneManifestIndex,
or LCA_Database object.
Use `CollectionManifest.load_from_filename(...)` to load the manifest
directly as a manifest object.

Implementation Details
----------------------

SqliteIndex:
* Hashes with values above MAX_SQLITE_INT=2**63-1 are transformed into
  signed long longs upon insertion, and then back into ulong longs upon
  retrieval.
* Hash overlap is calculated via a SELECT.
* SqliteIndex relies on SqliteCollectionManifest for manifest functionality,
  including signature selection and picklists.
SqliteCollectionManifest:
* each object maintains info about whether it is being "managed" by a
  SqliteIndex class or not. If it is, `_insert_row(...)` cannot be
  called directly.
* `select(...)` operates directly with SQL queries, except for
  picklist selection, which involves inspect each manifest row in
  Python. In addition to being (much) simpler, this ends up being
  faster in some important real world situations, even for millions of
  rows!
* filter_on_rows and filter_on_columns also both operate in Python,
  not SQL.
* for this reason, the `locations()` method returns a superset of
  locations.  This is potentially very significant if you do a select
  with a picklist that ignores most sketches - the `locations()`
  method will ignore the picklist.
  
Limitations:
* all of these classes share a single connection object, and it could
  get confusing quickly if you simultaneously insert and query. We suggest
  separating creation and insertion. That having been said, these databases
  should work fine for many simultaneous queries; just don't write :).
```
TODO testing: test internal and command line for,
- [x] creating an index
- [x] loading an index
- [x] loading an index as a manifest (should work)
- [x] loading an index as a standalone manifest and inserting (should fail)
- [x] creating a manifest, using from CLI
- [x] loading a manifest as a standalone index
- [x] loading a manifest as a standalone index in wrong directory
- [x] loading a manifest as a standalone index and insert (should succeed)
- [x] loading/using a checked-in index
- [x] loading/using a checked-in manifest
- [x] loading a lineage db with old table
- [x] loading a lineage db with new table name/sourmash info
- [x] loading a checked-in lineage db with old table
- [x] loading a checked-in lineage db with new table
- [x] do some realistic benchmarking of SqliteIndex and LCA_Database
- [x] check LCA database is loaded by load_sqlite_index.
- [x] check 'summarize' output on all three
- [x] test identifiers with '.' in sql lca dbs...
- [X] test LCA_SqliteDatabase.open with a SqliteIndex.
- [X] test LCA_SqliteDatabase.open with an empty SqliteIndex
- [x] implement update lca DB/on disk insert stuff for lca index?
- [x] test LCA_SqliteDatabase with a different lineagedb class?
- [X] add `-F sql` to sig check for manifest output
- [x] test `SaveSignatures` SQLite class repr.
- [x] test `db_outfile += '.lca.json'` line in lca_db.py
- [x] test `raise Exception(f"unknown save format for LCA_Database: '{format}'")` in lca_db.py
- [x] test `append= True` in manifest.py
- [x] test `__eq__` in CollectionManifest where manifests are NOT equal`p
- [x] test sqlite_utils/bad version code
- [x] test tax_utils/bad version code
- [x] test tax_utils/no version code
- [x] Write more about new `sourmash_internal` SQLite table, and also how the various sqlite things are compatible.
- [x] document new taxonomy foo in PR description

### CODE REVIEW

**Comment by ctb** [comment]
I think this is ready for review @sourmash-bio/devs! 🎉 

(What's another 3,000 lines of code between friends, anyway?)

File: doc/command-line.md [hunk 1]
```diff
@@ -931,6 +931,15 @@ As of sourmash 4.2.0, `lca index` supports `--picklist`, to
...
 exclude a few signatures from an index being built from a large collection.
 
+As of sourmash 4.4.0, `lca index` can produce an _on disk_ LCA
+database using SQLite. To prepare such a database, use
+`sourmash lca index ... -F sql`.
+
+All sourmash commands work with either type of LCA database (the
+default JSON database, and the SQLite version). SQLite databases are
+larger than JSON databases on disk but are typically much faster and
+to load and search, and use much less memory.
```
Comment (mr-eyes) [inline doc/command-line.md]:
```suggestion
larger than JSON databases on disk, but are typically much faster to load and search, and use much less memory.
```

Comment (ctb) [reply doc/command-line.md]: thanks! fixed in c05d044f8e44062

**Comment by ctb** [comment]
oops - I think I found a problem! `detect_types=PARSE_DECLTYPES` isn't needed! I'll verify later. (No need to hold review, removing it can only improve performance.)

**Comment by mr-eyes** [comment]
I am pulling updates from `latest` to see how #1976 will affect the tests on gh-actions.

File: src/sourmash/index/sqlite_index.py [hunk 2]
```diff
@@ -0,0 +1,1087 @@
...
+
+    @classmethod
+    def _create_tables(cls, cursor):
+        "Create the manifest table."
+        # this is a class method so that it can be used by SqliteIndex to
+        # create manifest-compatible tables.
+
+        sqlite_utils.add_sourmash_internal(cursor, 'SqliteManifest', '1.0')
+        cursor.execute("""
+        CREATE TABLE sourmash_sketches
```
Comment (mr-eyes) [inline src/sourmash/index/sqlite_index.py:632]:
```suggestion
        CREATE TABLE IF NOT EXISTS sourmash_sketches
```

File: src/sourmash/tax/tax_utils.py [hunk 3]
```diff
@@ -831,7 +879,7 @@ class TEXT,
             raise ValueError(f"taxonomy table already exists in '{filename}'")
 
         # follow up and create index
-        cursor.execute("CREATE UNIQUE INDEX taxonomy_ident ON taxonomy(ident);")
+        cursor.execute("CREATE UNIQUE INDEX sourmash_taxonomy_ident ON sourmash_taxonomy(ident);")
```
Comment (mr-eyes) [inline src/sourmash/tax/tax_utils.py:882]:
```suggestion
        cursor.execute("CREATE UNIQUE INDEX IF NOT EXISTS sourmash_taxonomy_ident ON sourmash_taxonomy(ident);")
```

Comment (mr-eyes) [reply src/sourmash/tax/tax_utils.py:882]: Running `sourmash lca index ....` twice with the same parameters will try to open and create tables and then threw exceptions. The failed table creation exception was handled in https://github.com/sourmash-bio/sourmash/blob/9e1ea60626a297c82c868e748ab7b0dcc16b8269/src/sourmash/tax/tax_utils.py#L879 which will prevent the process to complete. 

If `CREATE TABLE IF EXISTS` was added to that line, running `lca index` will open and update the same sqlite file and that will change the file content from the first run with now any thrown errors. 

**TL;DR** I think it would be better if `lca index` can detect that there's an already saved sqlite database and no need to re-run the command.

**Comment by mr-eyes** [comment]
Is building SBT SQLite index supported in the command line? 
`sourmash index` does not contain `-F {json,sql}, --database-format {json,sql}`.

File: src/sourmash/lca/lca_db.py [hunk 4]
```diff
@@ -104,7 +119,11 @@ def _get_ident_index(self, ident, fail_on_duplicate=False):
         return idx
 
     def _get_lineage_id(self, lineage):
-        "Get (create if nec) a unique lineage ID for each LineagePair tuples."
+        """Get (create if nec) a unique lineage ID for each
```
Comment (mr-eyes) [inline src/sourmash/lca/lca_db.py]:
```suggestion
        """Get (create if necessary) a unique lineage ID for each
```

**Comment by ctb** [comment]
> Is building SBT SQLite index supported in the command line? `sourmash index does not contain `-F {json,sql}, --database-format {json,sql}`.

`sourmash index` only builds SBTs; we've got an issue about changing that (https://github.com/sourmash-bio/sourmash/issues/949). 

SQLite databases (unlike SBT or LCA indexes) can be created with `sourmash sig cat ... -o xyz.sqldb`.

I agree this is confusing and inconsistent 😆 . I'll think about what to do - at the very least, better docs would be a good idea!

**Comment by ctb** [comment]
> > Is building SBT SQLite index supported in the command line? `sourmash index does not contain `-F {json,sql}, --database-format {json,sql}`.

> I agree this is confusing and inconsistent 😆 . I'll think about what to do - at the very least, better docs would be a good idea!

updated docs in 16b8b9b7, and added some suggestions in #1949.

File: src/sourmash/tax/tax_utils.py (hunk 3 above)
Comment (ctb) [reply src/sourmash/tax/tax_utils.py:882]: I think for SQL databases, you are 100% right :). But `sourmash lca index` needs to handle JSON databases too, and there are also some issues around the logic for appending that would not be handled well in the JSON case and might require a lot of extra checks.

I ended up going a different route in c443ae10 -
* we now forbid overwriting / appending to existing LCA database;
* I added command-line and API tests for LCA database saving x 2 (all named `twice`);

File: src/sourmash/index/sqlite_index.py (hunk 2 above)
Comment (ctb) [reply src/sourmash/index/sqlite_index.py:632]: Yep, this was intentional; the append case is tested and handled elsewhere, so I wanted an error here!

**Comment by ctb** [comment]
Thanks for catching the `lca index` problems with running twice!

**Comment by mr-eyes** [comment]
There are inconsistent results when running `lca classify` on same lca database built with both `SQL` and `json`. You can reproduce here:

```bash
curl -L https://osf.io/bw8d7/download?version=1 -o delmont-subsample-sigs.tar.gz
tar xzf delmont-subsample-sigs.tar.gz
curl -O -L https://github.com/ctb/2017-sourmash-lca/raw/master/tara-delmont-SuppTable3.csv
sourmash lca index -f tara-delmont-SuppTable3.csv delmont.lca.json delmont-subsample-sigs/*.sig
sourmash lca index -F sql -f tara-delmont-SuppTable3.csv delmont.lca.sql delmont-subsample-sigs/*.sig
sourmash lca classify --db delmont.lca.json --query delmont-subsample-sigs/TARA_RED_MAG_00003.fa.gz.sig > json_results
sourmash lca classify --db delmont.lca.sql --query delmont-subsample-sigs/TARA_RED_MAG_00003.fa.gz.sig > sql_results
diff json_results sql_results
```

Using the SQLite database gave no matches, while there is a match using the JSON version.

**Comment by ctb** [comment]
> There are inconsistent results when running `lca classify` on same lca database built with both `SQL` and `json`.

WOW, very nice catch!

I gotta say I was pretty worried when I first saw this, but it turns out that rather than being some deep and structural flaw in how I was handling identifiers differently b/t SQL and JSON, it was just a straight up simple programming error 😅.

Fixed in 52bc90ba.

**Comment by mr-eyes** [comment]
> WOW, very nice catch!

Thank you! 

> I gotta say I was pretty worried when I first saw this.

Likewise! and was resisting myself to debugging xD Glad it was straightforward.

**Comment by ctb** [comment]
> Likewise! and was resisting myself to debugging xD Glad it was straightforward.

it took me longer to track it down than I expected since I was gearing up for a big debugging effort for a subtle effect 😆

**Review by mr-eyes — APPROVED** [review]
This was a quite big PR to review, maybe I missed things or complicated-heuristical bugs while reviewing but I did a lot of testing and code inspection that I feel confident to merge.

I think it's good to wait for merging #2000 into this PR before merging to `latest`.

**Comment by ctb** [comment]
> This was a quite big PR to review, maybe I missed things or complicated-heuristical bugs while reviewing but I did a lot of testing and code inspection that I feel confident to merge.

yay, thank you!

> I think it's good to wait for merging #2000 into this PR before merging to `latest`.

I'm going to merge it now, tho, since I haven't decided on the scope of #2000 yet; might be bigger than just a `convert` command!
